### ISSUE TITLE
Given Channel's flow control and lack of explicit thread safety, how do I run a command without deadlocking?

### ISSUE THREAD

**Issue by SyntaxColoring** [paramiko/paramiko#1790]
Suppose I want to:

1. Execute some arbitrary command, where I don't necessarily know anything about how it will use stdout and stderr.
2. Block until the command completes.
3. Return its status code as an `int`, stdout output as a `bytes`, and stderr output as a separate `bytes`.

`client.exec_command(command)` gives us separate `stdout` and `stderr` *file-like objects.* But, how do I safely read from them in a way that won't risk deadlocking?

Specifically, since a `Channel` seems to encapsulate both a stderr and stdout stream, I assume that both streams share flow control. So, I'm concerned about:

* `stdout.read()` hanging if there's too much data on stderr waiting to be read.
* `stderr.read()` hanging if there's too much data on stdout waiting to be read.

The unrelated [`recv_exit_status()` docs](http://docs.paramiko.org/en/latest/api/channel.html#paramiko.channel.Channel.recv_exit_status) suggest running `Channel.recv()` in the background to work around a separate but related deadlock hazard:

> In some situations, receiving remote output larger than the current `Transport` or session’s `window_size` (e.g. that set by the `default_window_size` kwarg for `Transport.__init__`) will cause `recv_exit_status` to hang indefinitely if it is called prior to a sufficiently large `Channel.recv` (or if there are no threads calling `Channel.recv` in the background).
> 
> In these cases, ensuring that `recv_exit_status` is called after `Channel.recv` (or, again, using threads) can avoid the hang.

Which *maybe* hints, indirectly, that the expected way to accomplish this is something like this:

```python
stdout = None
stderr = None

def read_stdout():
    nonlocal stdout
    stdout = client.read()
stdout_thread = threading.Thread(target=read_stdout)

def read_stderr():
    nonlocal stderr
    stderr = client.read()
stderr_thread = threading.Thread(target=read_stderr)

stdout_thread.start()
stderr_thread.start()
stdout_thread.join()
stderr_thread.join()

status = stdout.channel.recv_exit_status()
```

But, other than that `recv_exit_status()` warning, it doesn't seem like the docs mention threading at all. There's no explicit guarantee that it's safe to `read()` from complementary stdout/stderr streams from multiple threads. And assuming thread safety seems like a bad idea.

What am I missing?

**Comment by ploxiln** [comment]
Threads are indeed the way to go here, it was pretty much designed for threads. Fabric-1.x does this:
https://github.com/fabric/fabric/blob/1.14/fabric/operations.py#L794-L832

**Comment by SyntaxColoring** [comment]
That's good to hear.

Would it be appropriate to add a note like this to the documentation of `recv()` and `recv_stderr()`?

> You can safely call `recv()` and `recv_stderr()` concurrently from separate threads.

Appropriately modified to include any other functions that are also part of the guarantee, of course.

(When you say "it was pretty much designed for threads," I assume that the "it" you're referring to is this particular set of functions, rather than Paramiko overall. [This random Stack Overflow answer](https://stackoverflow.com/a/50421992/497934) says that "Paramiko is not thread-safe.")

**Comment by ploxiln** [comment]
Yeah, that's true, most things you shouldn't do to the same SSHClient or Transport from different threads concurrently. But for stdin/stdout/stderr of a channel (including the file-like objects returned from `SSHClient.exec_command()`), it seems pretty common to have separate threads for each. That documentation update seems reasonable.

**Comment by pkittenis** [comment]
Race conditions beware. There is no guarantee
```
stdout_thread.join()
stderr_thread.join()
```
will not dead lock. If you write to stderr only, `channel.recv()` never returns and code is blocked on `stdout_thread.join()`.

There will need to be substantial amount of client code to handle this correctly. Note that paramiko already uses its own threads.

**Comment by SyntaxColoring** [comment]
@pkittenis 

> If you write to stderr only, `channel.recv()` never returns and code is blocked on `stdout_thread.join()`.

Are you sure? Why would `channel.recv()` never return? I would expect it to return `b''` if the command completes without writing anything to stdout.

Here's a quick test I just ran:

```python
>>> stdin, stdout, stderr = c.exec_command('dd if=/dev/urandom of=/dev/stderr bs=1k count=1024')
>>> stdout.read()
b''
>>> stderr.read()
b'[1 MB of random data]'

**Comment by pkittenis** [comment]
That's not been my experience with `makefile` streams at least. Nor has the behaviour been consistent, being race conditions.

**Comment by SyntaxColoring** [comment]
> See #1778 for another report of dead locking from reading stdout, among the multiple open issues for similar behaviour.

#1778 does this:

```python
(stdin, stdout, stderr) = client.exec_command(command, timeout=5)

stdout_data = stdout.read() # <= blocking and timeout ... !!
stderr_data = stderr.read()
```

I think the problem in this example is that `stdout.read()` blocks because too much *stderr* data has accumulated, and it's completely filling up the receive buffer. That stderr data needs to be read to make room for the remaining stdout data, so `stdout.read()` can finally see EOF and return. But `stderr.read()` can only begin once `stdout.read()` returns...hence, deadlock. The coding error is in running `stdout.read()` and `stderr.read()` *sequentially.*

This is precisely the problem that the threading in my proposed solution is supposed to work around. It runs `stdout.read()` and `stderr.read()` *concurrently* so neither one can block the other. At least, that's the idea.

Does that make sense, or does my solution still seem racy, @pkittenis?
